### ISSUE TITLE
[java] SingletonClassReturningNewInstance false positive with double assignment

### ISSUE THREAD

**Issue by alixwar** [pmd/pmd#932]
**Affects PMD Version:** 
 5.4.2

**Rule:**
[SingletonClassReturningNewInstance](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#singletonclassreturningnewinstance)

**Description:**
The rule incorrectly indicates that the code sample always returns a new instance:

> getInstance method always creates a new object and hence does not comply to Singleton Design Pattern behaviour. 

**Code Sample demonstrating the issue:**

```java
public final class Foo {

    private static volatile Foo myFoo;

    public static Foo getInstance() {
        Foo instance = myFoo;
        if (instance == null) {
            synchronized (Foo .class) {
                instance = myFoo;
                if (instance == null) {
                    myFoo= instance = new Foo();
                }
            }
        }
        return instance;
    }

}
```

**Running PMD through:** *Gradle (SonarQube)*

**Comment by adangel** [comment]
@alixwar You're right - that's a false positive.
I would however slightly simplify the code using the pattern from [The "Double-Checked Locking is Broken" Declaration](http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html), avoid the local variable:

```java
public final class Foo {
    private static volatile Foo myFoo;

    public static Foo getInstance() {
        if (myFoo == null) {
            synchronized (Foo .class) {
                if (myFoo == null) {
                    instance = new Foo();
                }
            }
        }
        return myFoo;
    }
}
```

But since the singleton you are using, is static (there is only one Foo instance), there is an even simpler way:

```java
public final class Foo {
    private Foo() {}

    private static class FooSingletonHelper {
        private static final Foo INSTANCE = new Foo();
    }

    public static Foo getInstance() {
        return FooSingletonHelper.INSTANCE;
    }
}
```

See also [Java Singleton Design Pattern Best Practices with Examples](https://www.journaldev.com/1377/java-singleton-design-pattern-best-practices-examples)

**Comment by alixwar** [comment]
@adangel Thanks for the pointers! One detail: the Bill Pugh Singleton Implementation that you recommend triggers another rule violation:

{code} private static final Foo  INSTANCE = new  Foo();{code}
> Instantiates through a private constructor from outside of the constructors class (pmd:AccessorClassGeneration)

**Comment by liu78778** [comment]
> @adangel Thanks for the pointers! One detail: the Bill Pugh Singleton Implementation that you recommend triggers another rule violation:
> 
> `private static final Foo INSTANCE = new Foo();`
> 
> > Instantiates through a private constructor from outside of the constructors class (pmd:AccessorClassGeneration)

I encountered the same situation, when using a static inner class singleton.
(pmd:AccessorClassGeneration)

![1](https://user-images.githubusercontent.com/93829/55721187-45a94b00-5a35-11e9-81f2-7408f9bace6b.png)

**Comment by jsotuyod** [comment]
@alixwar @liu78778 long story short: the solution is to change the visibility of the constructor from `private` to package-private. It's slightly more open, but is what the compiler ends up doing anyway (through a synthetic constructor).

Now, having said that, I'd strongly recommend you revise wether you actually need the `AccessorClassGeneration` rule. The rule (along with the companion `AccessorMethodGeneration`) warn when you access a private field / method / constructor from an inner class (or viceversa). This access is impossible in Java, so the syntax sugar is translated by the compiler into:

1. creating an "accessor" method (typically named `access$001` and so forth) with package-private visibility, delegating to the private method / working as a setter / getter for the private field; or simply a package-private constructor delegating to the private one.
2. editing the caller to call the new accessor method instead of the original private one

The extra indirection introduced by this method has 2 negative effects:

1. a slightly worst performance (minimal, and typically removed by the JVM after warmup)
2. a greater number of methods in the class (something you should not even care about, except for some very specific situations, such as aiming for single-dexing an Android application).

I personally never adopt this rule outside of an Android project myself.

**Comment by liu78778** [comment]
@jsotuyod Ah, I get it, thank you very much for your replay
